File: engines/v8/ts/buffer.ts
import type { NativeV8Bindings } from "./engine"

export class V8Buffer {
  private native: NativeV8Bindings
  private contextHandle: number
  private handle: number

  constructor(native: NativeV8Bindings, contextHandle: number, sizeOrData?: number | string | Uint8Array, encoding?: string) {
    this.native = native
    this.contextHandle = contextHandle

    if (typeof sizeOrData === "number") {
      this.handle = native.valueNewArrayBuffer(contextHandle, sizeOrData)
    } else if (typeof sizeOrData === "string") {
      const enc = encoding ?? "utf8"
      if (enc === "hex") {
        this.handle = native.encodingHexDecode(contextHandle, sizeOrData)
      } else if (enc === "base64") {
        this.handle = native.encodingBase64Decode(contextHandle, sizeOrData)
      } else {
        const encoded = native.encodingEncode(contextHandle, sizeOrData)
        this.handle = encoded
      }
    } else if (sizeOrData instanceof Uint8Array) {
      this.handle = native.bufferFromBytes(contextHandle, sizeOrData, sizeOrData.length)
    } else {
      this.handle = native.valueNewArrayBuffer(contextHandle, 0)
    }
  }

  get length(): number {
    return this.native.bufferGetLength(this.contextHandle, this.handle)
  }

  toString(encoding?: string, start?: number, end?: number): string {
    const enc = encoding ?? "utf8"
    const s = start ?? 0
    const e = end ?? this.length
    const result = this.native.bufferToString(this.contextHandle, this.handle, enc, s, e)
    return result.data ?? ""
  }

  toJSON(): { type: string; data: number[] } {
    return { type: "Buffer", data: Array.from(this.toBytes()) }
  }

  toBytes(): Uint8Array {
    const ptr = this.native.bufferGetData(this.contextHandle, this.handle)
    const len = this.length
    const arr = new Uint8Array(len)
    for (let i = 0; i < len; i++) arr[i] = this.native.getValueAtIndex(this.contextHandle, this.handle, i)
    return arr
  }

  slice(start: number, end?: number): V8Buffer {
    const e = end ?? this.length
    const handle = this.native.bufferSlice(this.contextHandle, this.handle, start, e)
    const buf = Object.create(V8Buffer.prototype) as V8Buffer
    buf.native = this.native
    buf.contextHandle = this.contextHandle
    buf.handle = handle
    return buf
  }

  copy(target: V8Buffer, targetStart?: number, sourceStart?: number, sourceEnd?: number): number {
    const ts = targetStart ?? 0
    const ss = sourceStart ?? 0
    const se = sourceEnd ?? this.length
    const count = se - ss
    const result = this.native.bufferCopy(this.contextHandle, target.handle, ts, this.handle, ss, count)
    return result.success ? count : 0
  }

  static concat(list: V8Buffer[], totalLength?: number): V8Buffer {
    if (list.length === 0) throw new Error("empty list")
    const first = list[0]
    const handles = list.map(b => b.handle)
    const handle = first.native.bufferConcat(first.contextHandle, handles, handles.length)
    const buf = Object.create(V8Buffer.prototype) as V8Buffer
    buf.native = first.native
    buf.contextHandle = first.contextHandle
    buf.handle = handle
    return buf
  }

  static from(data: string | Uint8Array | number[], encoding?: string): V8Buffer {
    if (typeof data === "string") return new V8Buffer(undefined as unknown as NativeV8Bindings, 0, data, encoding)
    if (data instanceof Uint8Array) return new V8Buffer(undefined as unknown as NativeV8Bindings, 0, data)
    return new V8Buffer(undefined as unknown as NativeV8Bindings, 0, new Uint8Array(data))
  }

  static alloc(size: number, fill?: number): V8Buffer {
    const buf = new V8Buffer(undefined as unknown as NativeV8Bindings, 0, size)
    if (fill !== undefined) buf.fill(fill)
    return buf
  }

  fill(value: number, offset?: number, end?: number): this {
    const o = offset ?? 0
    const e = end ?? this.length
    for (let i = o; i < e; i++) {
      this.native.setValueAtIndex(this.contextHandle, this.handle, i, value)
    }
    return this
  }

  write(data: string, offset?: number): number {
    const o = offset ?? 0
    const encoded = this.native.encodingEncode(this.contextHandle, data)
    const len = this.native.bufferGetLength(this.contextHandle, encoded)
    this.native.bufferCopy(this.contextHandle, this.handle, o, encoded, 0, len)
    return len
  }

  dispose(): void {
    this.native.valueDispose(this.handle)
  }
}

File: engines/v8/src/buffer.rs
use crate::error::V8Error;

#[cfg(feature = "native")]
use crate::ffi;
#[cfg(feature = "native")]
use std::ffi::CString;

pub struct V8Buffer {
    #[cfg(feature = "native")]
    ctx: *mut ffi::V8ContextHandle,
    #[cfg(feature = "native")]
    handle: *mut ffi::V8ValueHandle,
    #[cfg(feature = "native")]
    owns: bool,
}

impl V8Buffer {
    #[cfg(feature = "native")]
    pub fn new(ctx: *mut ffi::V8ContextHandle, size: usize) -> Result<Self, V8Error> {
        let handle = unsafe { ffi::klyron_v8_buffer_new(ctx, size) };
        if handle.is_null() { Err(V8Error::EvalFailed("buffer_new failed".into())) }
        else { Ok(Self { ctx, handle, owns: true }) }
    }

    #[cfg(not(feature = "native"))]
    pub fn new(_ctx: *mut std::ffi::c_void, _size: usize) -> Result<Self, V8Error> {
        Err(V8Error::NotInitialized)
    }

    #[cfg(feature = "native")]
    pub fn from_string(ctx: *mut ffi::V8ContextHandle, s: &str, encoding: Option<&str>) -> Result<Self, V8Error> {
        let c_str = CString::new(s).map_err(|e| V8Error::EvalFailed(e.to_string()))?;
        let c_enc = encoding.map(|e| CString::new(e)).transpose().map_err(|e| V8Error::EvalFailed(e.to_string()))?;
        let handle = unsafe {
            ffi::klyron_v8_buffer_from_string(ctx, c_str.as_ptr(),
                                               c_enc.as_ref().map_or(std::ptr::null(), |e| e.as_ptr()))
        };
        if handle.is_null() { Err(V8Error::EvalFailed("buffer_from_string failed".into())) }
        else { Ok(Self { ctx, handle, owns: true }) }
    }

    #[cfg(feature = "native")]
    pub fn from_bytes(ctx: *mut ffi::V8ContextHandle, data: &[u8]) -> Result<Self, V8Error> {
        let handle = unsafe { ffi::klyron_v8_buffer_from_bytes(ctx, data.as_ptr(), data.len()) };
        if handle.is_null() { Err(V8Error::EvalFailed("buffer_from_bytes failed".into())) }
        else { Ok(Self { ctx, handle, owns: true }) }
    }

    #[cfg(feature = "native")]
    pub fn handle(&self) -> *mut ffi::V8ValueHandle {
        self.handle
    }

    #[cfg(feature = "native")]
    pub fn length(&self) -> usize {
        unsafe { ffi::klyron_v8_buffer_get_length(self.ctx, self.handle) }
    }

    #[cfg(feature = "native")]
    pub fn to_string(&self, encoding: Option<&str>, start: Option<usize>, end: Option<usize>) -> Result<String, V8Error> {
        let c_enc = encoding.map(|e| CString::new(e)).transpose().map_err(|e| V8Error::EvalFailed(e.to_string()))?;
        let r = unsafe {
            ffi::klyron_v8_buffer_to_string(self.ctx, self.handle,
                                             c_enc.as_ref().map_or(std::ptr::null(), |e| e.as_ptr()),
                                             start.unwrap_or(0), end.unwrap_or(self.length()))
        };
        if r.success {
            let s = if r.data.is_null() { String::new() }
            else {
                let s = unsafe { std::ffi::CStr::from_ptr(r.data).to_string_lossy().into() };
                unsafe { ffi::klyron_v8_free_string(r.data) };
                s
            };
            Ok(s)
        } else {
            Err(V8Error::EvalFailed("buffer_to_string failed".into()))
        }
    }

    #[cfg(feature = "native")]
    pub fn slice(&self, start: usize, end: usize) -> Result<Self, V8Error> {
        let handle = unsafe { ffi::klyron_v8_buffer_slice(self.ctx, self.handle, start, end) };
        if handle.is_null() { Err(V8Error::EvalFailed("buffer_slice failed".into())) }
        else { Ok(Self { ctx: self.ctx, handle, owns: true }) }
    }

    #[cfg(feature = "native")]
    pub fn copy_to(&self, target: &V8Buffer, dst_offset: usize, src_offset: usize, count: usize) -> Result<(), V8Error> {
        let r = unsafe { ffi::klyron_v8_buffer_copy(self.ctx, target.handle, dst_offset, self.handle, src_offset, count) };
        if r.success { Ok(()) }
        else { Err(V8Error::EvalFailed("buffer_copy failed".into())) }
    }

    #[cfg(feature = "native")]
    pub fn write(&self, data: &[u8], offset: usize) -> Result<(), V8Error> {
        let r = unsafe { ffi::klyron_v8_buffer_write(self.ctx, self.handle, data.as_ptr(), offset, data.len()) };
        if r.success { Ok(()) }
        else { Err(V8Error::EvalFailed("buffer_write failed".into())) }
    }
}

#[cfg(feature = "native")]
impl Drop for V8Buffer {
    fn drop(&mut self) {
        if self.owns && !self.handle.is_null() {
            unsafe { ffi::klyron_v8_value_dispose(self.handle) }
        }
    }
}
